### PR TITLE
added some change to the adapter

### PR DESCRIPTION
**Pull request by 0xantman**

### CODE REVIEW

File: src/contract.rs [hunk 1]
```diff
@@ -65,18 +65,29 @@ pub fn query<S: Storage, A: Api, Q: Querier>(
         QueryMsg::Holder { address } => to_binary(&query_holder(&deps, address)?),
         QueryMsg::Holders { start_after, limit } => {
             to_binary(&query_holders(&deps, start_after, limit)?)
-        }
+        },
+        QueryMsg::TransferFrom { .. } => to_binary(&query_transfer_from(deps)?),
     }
 }
 
+fn query_transfer_from<S: Storage, A: Api, Q: Querier>(
```
Comment (orkunkl) [inline src/contract.rs:73]: `TransferFrom` is an execution function and not related to querying, Also why you need TransferFrom in this contract anyways. I mean the transfer operations are in staking cw20 contract?

File: src/contract.rs [hunk 2]
```diff
@@ -65,18 +65,29 @@ pub fn query<S: Storage, A: Api, Q: Querier>(
...
+}
+
 fn query_config<S: Storage, A: Api, Q: Querier>(
     deps: &Extern<S, A, Q>,
 ) -> StdResult<ConfigResponse> {
     let config: Config = read_config(&deps.storage)?;
     Ok(ConfigResponse {
-        lottery_contract: deps.api.human_address(&config.lottery_contract)?,
-        cw20_token_contract: deps.api.human_address(&config.cw20_token_contract)?,
+        admin: deps.api.human_address(&config.admin)?,
```
Comment (orkunkl) [inline src/contract.rs:87]: Why removed external lottery contract address?

File: src/claim.rs [hunk 3]
```diff
@@ -1,11 +1,10 @@
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 
-use cosmwasm_std::{Addr, BlockInfo, Deps, StdResult, Storage, Uint128, CanonicalAddr};
-use cw0::Expiration;
-use cw_storage_plus::Map;
-use cw20::Expiration;
+use cosmwasm_std::{BlockInfo, StdResult, Storage, Uint128, CanonicalAddr};
+// use cw_storage_plus::Map;
```
Comment (0xantman) [inline src/claim.rs:5]: I commented this line since it seem to be not used

File: src/user.rs [hunk 4]
```diff
@@ -78,20 +75,53 @@ pub fn handle_claim_rewards<S: Storage, A: Api, Q: Querier>(
...
+    let address_raw = deps.api.canonical_address(&env.message.sender)?;
     let sender = env.message.sender;
-
-    // Check sender is token contract
-    if sender != deps.api.human_address(&config.cw20_token_contract)? {
-        return Err(StdError::unauthorized());
+    /*
+        TODO: Add safe_lock
+     */
+    /*if state.safe_lock {
```
Comment (orkunkl) [inline src/user.rs:101]: If you want to add safe lock, renounce function should be added

File: src/user.rs [hunk 5]
```diff
@@ -107,10 +137,10 @@ pub fn handle_increase_balance<S: Storage, A: Api, Q: Querier>(
     store_holder(&mut deps.storage, &address_raw, &holder)?;
     store_state(&mut deps.storage, &state)?;
     let res = HandleResponse {
-        messages: vec![],
+        messages: vec![msg],
         log: vec![
-            log("action", "increase_balance"),
-            log("holder_address", address),
+            log("action", "bond-lota-stake"),
```
Comment (orkunkl) [inline src/user.rs:142]: in ecnoded cosmos msgs and events the format is camel case

File: src/user.rs [hunk 6]
```diff
@@ -159,8 +200,8 @@ pub fn handle_unbound<S: Storage, A: Api, Q: Querier>(
     let res = HandleResponse {
         messages: vec![],
         log: vec![
-            log("action", "unbond_stake"),
-            log("holder_address", address),
+            log("action", "unbond_lota_stake"),
```
Comment (orkunkl) [inline src/user.rs:203]:
```suggestion
            log("action", "unbond_stake"),
```

File: src/user.rs [hunk 7]
```diff
@@ -78,20 +75,53 @@ pub fn handle_claim_rewards<S: Storage, A: Api, Q: Querier>(
...
+    }*/
+
+    if !env.message.sent_funds.is_empty() {
+        return Err(StdError::generic_err("Do not send funds with stake"));
     }
+    if amount.is_zero() {
+        return Err(StdError::generic_err("Amount required"));
+    }
+    // Prepare msg to send
+    let prepare_msg = QueryMsg::TransferFrom {
```
Comment (orkunkl) [inline src/user.rs:114]: Import and use HandleMsg::TransferFrom from cosmwasm-plus v0.2.3 https://github.com/CosmWasm/cosmwasm-plus/blob/v0.2.3/contracts/cw20-base/src/msg.rs#L101

File: src/user.rs [hunk 8]
```diff
@@ -78,20 +75,53 @@ pub fn handle_claim_rewards<S: Storage, A: Api, Q: Querier>(
     })
 }
 
-pub fn handle_increase_balance<S: Storage, A: Api, Q: Querier>(
+/*
+    Encode and prepare message to perfom a TransferFrom cw-20 LOTA contract in order to transfer
+    staker LOTA to staking contract funds are locked as a custodian
+*/
+fn encode_msg_execute(msg: QueryMsg, address: HumanAddr) -> StdResult<CosmosMsg> {
```
Comment (orkunkl) [inline src/user.rs:82]: create an helpers.rs file and move this code there

File: src/testing/tests.rs [hunk 9]
```diff
@@ -58,9 +59,10 @@ fn proper_init() {
     assert_eq!(
         config_response,
         ConfigResponse {
-            lottery_contract: HumanAddr::from(MOCK_HUB_CONTRACT_ADDR),
-            cw20_token_contract: HumanAddr::from(MOCK_CW20_CONTRACT_ADDR),
+            admin: HumanAddr::from(MOCK_HUB_CONTRACT_ADDR),
```
Comment (orkunkl) [inline src/testing/tests.rs:62]: Again why removed loterry contract?
The design I thought: when a round finishes lotterry contract sends jackpot tax reward to reward contract and then triggers update global index

Comment (orkunkl) [reply src/testing/tests.rs:62]: I would not make this admin

File: src/global.rs [hunk 10]
```diff
@@ -6,7 +6,9 @@ use cosmwasm_std::{
     Storage,
 };
 use terra_cosmwasm::TerraMsgWrapper;
-
+/*
+    TODO: need to remove or adapt this
```
Comment (orkunkl) [inline src/global.rs:10]: remove what?

**Review by orkunkl — CHANGES_REQUESTED** [review]
Added review

File: src/testing/tests.rs (hunk 9 above)
Comment (orkunkl) [reply src/testing/tests.rs:62]: If this is for multiple game contracts connecting to reward contract, then changing it to address is a good idea.

File: src/contract.rs (hunk 2 above)
Comment (0xantman) [reply src/contract.rs:87]: There is no need on the staking contract of the lottery address

File: src/contract.rs (hunk 1 above)
Comment (0xantman) [reply src/contract.rs:73]: We need allowance in order to move tokens from an user to our staking contract or you have another solution ? 
We still need a struct to encode the msg

File: src/user.rs (hunk 4 above)
Comment (0xantman) [reply src/user.rs:101]: Yes this is just to remember

File: src/contract.rs (hunk 1 above)
Comment (orkunkl) [reply src/contract.rs:73]: https://github.com/LoTerra/loterra-staking-contract/pull/1#discussion_r618604396

**Review by orkunkl — APPROVED** [review]

**Pull request merged**